Docs: added concat & concatAssumeInjective (#7718)
ClickHouse/ClickHouse: docs/en/query_language/functions/string_functions.md

└───────────────────────────────────┘
```

## concat {#concat}

Concatenates the strings listed in the arguments, without a separator.

**Syntax** 

```sql
concat(s1, s2, ...)
```

> Alias: `CONCAT`.

Reviewer: It's just case-insensitive:
```
SELECT cOncat('aa', 'bb')
FORMAT TSV

aabb
```

While the real alias is `||` binary operator which is part of SQL standard:
```
:) select 'aa' || 'bb' format TSV;

SELECT concat('aa', 'bb')
FORMAT TSV

aabb
```

**Parameters**

> - `si` – string or key. The key value can be any [data type](https://clickhouse.yandex/docs/en/data_types/).

Reviewer: * internal links should point to markdown files instead of where they are published
* and supporting any data type sounds incorrect:
```
SELECT concat('aa', 1)
FORMAT TSV

Received exception from server (version 19.8.3):
Code: 386. DB::Exception: Received from localhost:9000, ::1. DB::Exception: There is no supertype for types String, UInt8 because some of them are String/FixedString and some of them are not.

0 rows in set. Elapsed: 0.001 sec.
```

**Returned values**

Returns the string that results from concatenating the arguments. 

> If either argument is null, `concat` returns null. 

Reviewer: ```suggestion
If any of argument values is `NULL`, `concat` returns `NULL`. 
```

> Type: `String`.

Reviewer: Not always:
```
SELECT toTypeName(concat('aa', NULL))
FORMAT TSV

Nullable(Nothing)
```
or
```
SELECT toTypeName(concat(CAST(number, 'String'), if(number % 2, '', NULL)))
FROM numbers(1)
FORMAT TSV

Nullable(String)
```

**Example**

Input table:

```sql

> CREATE TABLE key_val(`key1` String, `key2` String, `value` UInt32) ENGINE = TinyLog

Reviewer: this is way too complicated example for this function, using just literals should be enough

INSERT INTO key_val VALUES ('Hello, ','World',1)('Hello, ','World',2)('Hello, ','World!',3)('Hello',', World!',2)
SELECT * from key_val
```

```text
┌─key1────┬─key2─────┬─value─┐
│ Hello,  │ World    │     1 │
│ Hello,  │ World    │     2 │
│ Hello,  │ World!   │     3 │
│ Hello   │ , World! │     2 │
└─────────┴──────────┴───────┘
```

Query:

```sql
SELECT concat(key1, key2), sum(value) FROM key_val GROUP BY concat(key1, key2)
```

Result:

```text
┌─concat(key1, key2)─┬─sum(value)─┐
│ Hello, World!      │          5 │
│ Hello, World       │          3 │
└────────────────────┴────────────┘
```

## concatAssumeInjective {#concatassumeinjective}

> Same as [concat](#concat), the difference is that you need to ensure that concat(s1, s2, ...) → sn is injective, it will be used for optimization of GROUP BY.

Reviewer: ```suggestion
Same as [concat](#concat), the difference is that you need to ensure that `concat(s1, s2, ...) → sn` is injective, it will be used for optimization of GROUP BY.
```
also we'd better link to GROUP BY chapter

**Syntax** 

```sql
concatAssumeInjective(s1, s2, ...)
```

**Parameters**

- `si` – can be any [data type](https://clickhouse.yandex/docs/en/data_types/).

**Returned values**

> Returns the string that results from concatenating the arguments after using GROUP BY.

Reviewer: what happens if it's called outside of `GROUP BY`?

└───────────────────────────────────┘
```

## concat {#concat}

Concatenates the strings listed in the arguments, without a separator.

**Syntax** 

```sql
concat(s1, s2, ...)
```

**Parameters**

> - `si` – string or key.

Reviewer 2: What does it mean?

**Returned values**

Returns the string that results from concatenating the arguments. 

If any of argument values is `NULL`, `concat` returns `NULL`. 

**Example**

Query:

```sql
SELECT concat('Hello, ', 'World!')
```

Result:

```text
┌─concat('Hello, ', 'World!')─┐
│ Hello, World!               │
└─────────────────────────────┘
```

## concatAssumeInjective {#concatassumeinjective}

Same as [concat](#concat), the difference is that you need to ensure that `concat(s1, s2, ...) → sn` is injective, it will be used for optimization of GROUP BY.

**Syntax** 

```sql
concatAssumeInjective(s1, s2, ...)
```

**Parameters**

- `si` – string or key.

**Returned values**

> Returns the string that results from concatenating the arguments after using GROUP BY.

Reviewer 2: Do not understand.
